### ISSUE TITLE
ERROR in f5c eventalign

### ISSUE THREAD

**Issue by loganylchen** [hasindu2008/f5c#121]
Hi there,

When I ran the following command,
```
f5c eventalign -r pass.fq.gz -b NATIVE_UCSC.bam -g transcriptome.fa --signal-index --scale-events -t 20 --rna -o NATIVE_UCSC_xpore.tsv
```
I've encountered some errors and the program abort.

Here is the log
```
[init_core::INFO] builtin RNA nucleotide model loaded
[E::fai_retrieve] Failed to retrieve block: unexpected end of file
[E::fai_retrieve] Failed to retrieve block: unexpected end of file
[E::fai_retrieve] Failed to retrieve block: unexpected end of file
[E::fai_retrieve] Failed to retrieve block: unexpected end of file 
...
[E::fai_retrieve] Failed to retrieve block: unexpected end of file
[E::fai_retrieve] Failed to retrieve block: unexpected end of file
[meth_main::112.482*0.10] 512 Entries (0.9M bases) loaded
f5c: src/eventalign.c:1110: std::string disambiguate(const string&): Assertion `isValid(out[i])' failed.
f5c: src/eventalign.c:1110: std::string disambiguate(const string&): Assertion `isValid(out[i])' failed.
[1]    60364 abort      f5c eventalign -r pass.fq.gz -b  -g    -t 20
```

Thanks in advance,

Logan

**Comment by hasindu2008** [comment]
What was the f5c index command you ran? Did it succeed?

**Comment by loganylchen** [comment]
I think so, this is the log of index process.

```
[parse_index_options::WARNING] --iop is incompatible with sequencing summary files. Option --sequencing-summary-file will be ignored
[parse_index_options::INFO] Consider using --slow5 option for fast indexing, methylation calling and eventalignment. See f5c section under https://hasindu2008.github.io/slow5tools/workflows.html for an example.
[find_all_fast5] Looking for fast5 in workspace
[f5c_index_iop] 815 fast5 files found - took 0.004s
[f5c_index_iop] Spawning 10 I/O processes to circumvent HDF hell
[f5c_index_iop] Parallel indexing done - took 5060.984s
[f5c_index_merge] Indexing merging done - took 7.495s.
[readdb] num reads: 3250116, num reads with path to fast5: 3250116
[main] Version: 1.0
[main] CMD: f5c index -s sequencing_summary.txt -d fast5/workspace -t 20 --iop 10 pass.fq.gz
[main] Real time: 5156.874 sec; CPU time: 495.232 sec; Peak RAM: 0.896 GB

```

**Comment by loganylchen** [comment]
Hi @hasindu2008 ,

Do i need to upload some data for you to look into?

**Comment by hasindu2008** [comment]
At a glance, what you have done is correct.  How large is this dataset? It would be great if you could provide me with a way to reproduce this. 

Meanwhile, could you also see if there is a transcriptome.fa.fai file next to transcriptome.fa and if so could you first delete the transcriptome.fa.fai and try the f5c eventalign command again? It could be that somehow the fasta reference fai index file is somehow corrupted.

**Comment by loganylchen** [comment]
@hasindu2008 Thanks for your help. You did figure out the problem that caused the issue. I found my transcriptome.fa.fai was not consistent with transcriptome.fa. When I updated the fa.fai index file, `f5c eventalign` seems to work normally. 

![image](https://user-images.githubusercontent.com/25150929/215660019-ed805818-6da6-4435-a47f-a2d01e01e38a.png)

Thanks, I will close this issue.

**Comment by hasindu2008** [comment]
Cool. Also, I suggest converting your FAST5 to BLOW5 so that it doesn't become a processing bottleneck. Is this  the dataset at https://github.com/nanopore-wgs-consortium/NA12878/blob/master/RNA.md by any chance?

**Comment by loganylchen** [comment]
@hasindu2008 I agree with you of coverting the FAST5 to BLOW5, and I like BLOW5 very much. Here I still used the FAST5 since there is some basecalling information stored in it. I will use the BLOW5 if I do need the signal data solely. 

BTW, what do you mean by "Is this the dataset at https://github.com/nanopore-wgs-consortium/NA12878/blob/master/RNA.md by any chance?"

**Comment by hasindu2008** [comment]
Yeh, if you want to keep the basecalls in FAST5 that is fine. Basecalls are better stored in a FASTQ separately, but ONT kept on writing it to FAST5 and that made various software rely on it. Finally, ONT has realised that FASTQ is better stored separately and latest FAST5 does not store the FASTQ anymore. But indeed all those tools which relied on this cannot be changed.

Another thing is when you do f5c index it has to go through all the FAST5s, then during event align another time through the FAST5s - so going through the inefficient format twice. So my recommended pipeline is, to use slow5tools to convert the FAST5 temporarily (you can delete them after f5c eventalign) - that way only once we go through FAST5 and then both f5c index and f5c eventalign will be doing fast access to BLOW5 - see Methylation calling or eventalignment using f5c in
 https://hasindu2008.github.io/slow5tools/workflows.html.

In your command line, I noticed "NATIVE_UCSC.bam", so I was wondering if you were using the NA12878 RNA dataset from the WGS consortium. I recently converted the whole dataset and rebasecalled them with latest Guppy for something else, could share if you want.

**Comment by loganylchen** [comment]
@hasindu2008 It is really nice of you. Definitely yes, I agree with you and will follow your suggestions to build my own pipeline. There was only one mistake in which I meant the basecall information. I didn't mean the sequences but the signals to base mapping results from Guppy, which they stored in "Analysis" group in the basecalled fast5 files. I knew there were several tools that could do so, `tombo resquiggle`, `scrappie (python package)`, `nanopolish or f5c eventalign`, etc. they were all doing similar things. However, --if i were wrong, correct me-- except for `tombo` using its own way to do so, other tools adopt segmentation codes from the `scrappie`, which I thought was very out of date. So in my opinion, Guppy could provide better basecalling results, which means it does a better segmentation process. So I prefer to extract the Guppy signal-to-base results rather than using `tombo resquiggle` nor `f5c eventalign`. 

P.S. I think we have talked about it in another issue, in which I asked you if blow5 could store `move table` as well.

**Comment by hasindu2008** [comment]
About the accuracy of segmentation:
I am not sure which method is the best as I have not compared the accuracy among these different methods. But I think @skovaka has done an accuracy comparison with https://github.com/skovaka/UNCALLED/tree/uncalled4 but not sure if those findings have been published yet.  @hiruna72 has been looking into f5c resquiggle (https://hasindu2008.github.io/f5c/docs/output#resquiggle) vs Guppy move table. Move table (see below) is also not perfect as the stride size is always a multiple of 5 and thus the transitions are not always aligned.

<img width="632" alt="image" src="https://user-images.githubusercontent.com/12987163/216734035-3ccb3a48-4659-4ccf-9ef3-af7b8979838e.png">

About move table in BLOW5:
BLOW5 in theory can store the move table as an auxiliary tag. But in practice, I did not implement this feature in slow5tools as the move table is not really raw signal. And with updated basecallers move table would change. So my thinking was it is better to be stored Guppy in the FASTQ header or a separate file, to prevent the raw signal data from being convoluted. My opinion was, storing such intermediate data in the FAST5 was a bad idea from the beginning.

And finally, recently ONT seem to have realised this and the most recent Guppy no longer stores the move table in the FAST5s, instead, it is output as a tag in unaligned S/BAM format, which I commend them for doing.  Butter-eel dev branch [https://github.com/Psy-Fer/buttery-eel/tree/dev] can now generate these move tables into unaligned sam directly from BLOW5, you can give it a try, but be vigilant as it is still not extensively tested.

I have converted all the data to BLOW5 and also have basecalled with a Guppy 6.1.3. From BLOW5+buttery-eel basecalling is anyway quite fast and I can do with 6.4.2 if you want. Do you want the basecalls or BLOW5 files?

**Comment by hasindu2008** [comment]
@loganylchen also any preferred way to share this data?

**Comment by loganylchen** [comment]
I shall reopen this issue, I will close after done the discussion with you (I have missed your reply several times).

**Comment by loganylchen** [comment]
@hasindu2008 Thanks.

I am not sure which way you can provide it, can you give me some options? 
I knew the buttery-eel, as I followed you on the GitHub and saw you starred the repo. Really appreciate the nice repo that I could try combining SLOW5 and Guppy together.

**Comment by hasindu2008** [comment]
You can download the FASTQ from the temporary link [here](https://unsw-my.sharepoint.com/:u:/g/personal/z5136909_ad_unsw_edu_au/Ea2ZtMbglNlAvPAQJYBSoTEBtE02VNfYiNUlu5ouzE2urw?download=1). I can upload the BLOW5 similarly - do you want a single merged BLOW5 file or one BLOW5 file per run?

**Comment by loganylchen** [comment]
@hasindu2008 Thanks, I prefer a single merged BLOW5 file. I will download them when I get to the lab on Monday.

**Comment by hasindu2008** [comment]
You can download the merged BLOW5 from [here](https://cloudstor.aarnet.edu.au/sender/download.php?token=f1a76482-5071-4fe1-9968-12a912fc7eea&files_ids=18452785) and the BLOW5 index from [here](https://cloudstor.aarnet.edu.au/sender/download.php?token=f1a76482-5071-4fe1-9968-12a912fc7eea&files_ids=18452790). The link will be available till 30/04/2023.

**Comment by hasindu2008** [comment]
@loganylchen did you download them?

**Comment by loganylchen** [comment]
Sorry for forgetting to inform you about the downloading. I have tried downloading the files, index file works well, but the merged blow5 file always fails after a period of downloading. I have used the raw fast5 to do the analysis. Thank you so much for your kind help.

**Comment by hasindu2008** [comment]
No problem.

**Comment by loganylchen** [comment]
I have a further question about `f5c eventalign`, as I noticed the number of reads in the eventalign results was smaller than what I input, and I've seen the log as below:
<img width="935" alt="image" src="https://user-images.githubusercontent.com/25150929/222854872-3fc32cbf-1ff8-4923-b649-c5fc1f38e2bb.png">

in which there are some reads that are labeled as `qc failed` and `no alignment`.

But the bam file I used here have been filtered by the samtools with `-F0x2324 -q 10 --min-qlen 80`. So I thought all the reads in the bam file should be included in the final eventalign results. 

Could you give me some hints on the issue here?

I also include the command I used in the `eventalign` as below:
```
f5c eventalign -r data/Oligo3/fastq/pass.fq.gz --print-read-names --signal-index --scale-events --min-mapq 0 --min-recalib-events 50 --samples --rna -b results/alignments/Oligo3_filtered.bam -g oligo.fa -t 32 --slow5 results/blow5/Oligo3.blow5 -o results/eventalign/Oligo3_nanocompore.tsv
```

**Comment by hasindu2008** [comment]
@loganylchen how many alignments were there in the input BAM file? Is it 769? 
you should have 769-283-242 in your output. 

if there are secondary alignments in your BAM and you want them to be mapped, you can give the option --secondary=yes.

**Comment by loganylchen** [comment]
Yes, the total number is 769.

I don't think there are secondary alignments in my bam file, as seen below:
<img width="485" alt="image" src="https://user-images.githubusercontent.com/25150929/222856571-89772728-12c5-404b-863c-1f21add40ed9.png">

**Comment by hasindu2008** [comment]
qc failed means those reads has too many stays (>5 event segments per base) and thus were not of good quality for the output to be trusted. no alignment means, the event to reference alignment failed or the alignment score was not good enough to be called a good alignment. If you run nanopolish event align,  you should get the same number of reads as these cut-off values directly come from nanopolish event align. 

As RNA data is noisy and both ONT basecall models (that output the fastq) and the HMM model are not as good as that is for DNA. So for RNA, a lot reads have this fate. 

We could relax these cut-offs, but then the output may not be that good. If you do not have sufficient data and want to rescue more data, I could investigate. However, that na12878 dataset has very high total yield right?

**Comment by hasindu2008** [comment]
Also, after how much percentage did your download fail? I just put a test download and it seems that it is running smoothly for 16% so far. Did you by any chance try to use the web browser to download - for large files like this, browser download could be a bit problematic.

```
$ wget "https://cloudstor.aarnet.edu.au/sender/download.php?token=f1a76482-5071-4fe1-9968-12a912fc7eea&files_ids=18452785" -O a.blow5
--2023-03-04 10:37:58--  https://cloudstor.aarnet.edu.au/sender/download.php?token=f1a76482-5071-4fe1-9968-12a912fc7eea&files_ids=18452785
Resolving cloudstor.aarnet.edu.au (cloudstor.aarnet.edu.au)... 202.158.207.20
Connecting to cloudstor.aarnet.edu.au (cloudstor.aarnet.edu.au)|202.158.207.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 613347601024 (571G) [application/binary]
Saving to: 'a.blow5’

a.blow5                            16%[========>                                                  ]  93.38G  58.4MB/s    eta 2h 9m
```

**Comment by loganylchen** [comment]
> qc failed means those reads has too many stays (>5 event segments per base) and thus were not of good quality for the output to be trusted. no alignment means, the event to reference alignment failed or the alignment score was not good enough to be called a good alignment. If you run nanopolish event align, you should get the same number of reads as these cut-off values directly come from nanopolish event align.
> 
> As RNA data is noisy and both ONT basecall models (that output the fastq) and the HMM model are not as good as that is for DNA. So for RNA, a lot reads have this fate.
> 
> We could relax these cut-offs, but then the output may not be that good. If you do not have sufficient data and want to rescue more data, I could investigate. However, that na12878 dataset has very high total yield right?

I see. I think the settings are reasonable.

**Comment by loganylchen** [comment]
> Also, after how much percentage did your download fail? I just put a test download and it seems that it is running smoothly for 16% so far. Did you by any chance try to use the web browser to download - for large files like this, browser download could be a bit problematic.
> 
> ```
> $ wget "https://cloudstor.aarnet.edu.au/sender/download.php?token=f1a76482-5071-4fe1-9968-12a912fc7eea&files_ids=18452785" -O a.blow5
> --2023-03-04 10:37:58--  https://cloudstor.aarnet.edu.au/sender/download.php?token=f1a76482-5071-4fe1-9968-12a912fc7eea&files_ids=18452785
> Resolving cloudstor.aarnet.edu.au (cloudstor.aarnet.edu.au)... 202.158.207.20
> Connecting to cloudstor.aarnet.edu.au (cloudstor.aarnet.edu.au)|202.158.207.20|:443... connected.
> HTTP request sent, awaiting response... 200 OK
> Length: 613347601024 (571G) [application/binary]
> Saving to: 'a.blow5’
> 
> a.blow5                            16%[========>                                                  ]  93.38G  58.4MB/s    eta 2h 9m
> ```

Yes, I used the browser to download the file. I will try the wget instead this time. Thanks.

**Comment by loganylchen** [comment]
@hasindu2008 for downloading the blow5, I have tried several times but always encountered "closed connection" after downloading around 40G data. But I think it is OK, as I have downloaded the RAW FAST5 data and could basecall and covert them into BLOW5 by myself. Thanks. 

I think, for now, all my questions have been conquered.

**Comment by hasindu2008** [comment]
@loganylchen It appears that the aarnet filesender is broken. Conversion was a bit of a pain as these data were a mix of old multi and single fast5 of different conventions, so I guess it is easier to just download a converted verison.

I have uploaded it to an aws s3 bucket and you should be able to download it easily this time from https://slow5test.s3.amazonaws.com/tmp/na12878_rna_merged.blow5.

**Comment by hasindu2008** [comment]
@loganylchen 
Did this download link succeed?

**Comment by loganylchen** [comment]
@hasindu2008 Yes, just finished downloading after two days. Thanks

**Comment by hasindu2008** [comment]
Great, will delete it now. It will save you some time for manual FAST5 fixing. :D
